docs/src/adr/016adr-simple.md

```diff
@@ -0,0 +1,137 @@
+```tla
+CONSTANT Values
+VARIABLE messages
+
+SendT1(v) == messages \union { [ type |-> "t1", x |-> v ] }
+
+ReadT1 == 
+  \E msg \in messages:
+    /\ msg.type = "t1"
+    /\ F(msg.x) \* some action
+
+Next ==
+  \/ \E v \in Values: SendT1(v)
+  \/ ReadT1
+```
+
+The central object is the set `messages`, which is modified at each step, and contains records. This makes it one of the more expensive expressions, in terms of the underlying SMT encodings in Apalache. However, as far as the use of `messages` goes, there is a major insight to be had: _it is not necessary to model `messages` explicitly for `ReadT1`, we only need to specify the property that certain messages of type "t1" with given payloads have been sent_. Naturally, modeling `messages` explicitly is _sufficient_ for this purpose, but if one wanted to avoid the use of sets and records, one could write the following specification instead:
+```tla
+CONSTANT Values
+VARIABLE T1messages
+
+SendT1(v) == [T1messages EXCEPT ![v] = TRUE]
+
+ReadT1 ==
+  \E v \in Values:
+    /\ T1messages[v]
+    /\ F(v) \* some action
+
+Next ==
+  \/ \E v \in Values: SendT1(v)
+  \/ ReadT1
+```
+
+By encoding, for example, set membership checks as characteristic function evaluations, one can write some specifications in a fragment of TLA+ that avoids all complex data structures, sets-of-sets, records, sequences, and so on, and replaces them with predicates (functions).
+Rewriting specifications in this way is nontrivial, and shouldn't be expected of engineers, however, should a specification author undertake such a transformation, we should be able to provide some payoff.
+If we only limit ourselves to specifications in this restricted fragment (defined explicitly below), the current SMT encoding is needlessly complex. 
+We can implement a specialized encoding, which does not use arena logic of any kind, but is much more direct and even lends itself well to multiple kinds of solvers (e.g. IVy or VMT, in addition to standard SMT).
+
+## Options
+
+<!-- Communicate the options considered.
+     This records evidence of our circumspection and documents the various alternatives
+     considered but not adopted.
+-->
+1. Reuse most of the existing implementation and encoding, with a modified language watchdog, then output an SMT file from the context, instead of solving.
+    - Pros:
+
+      - Little work
+
+    - Cons:
+
+      - Locked to the SMT format
+      - Unnecessary additional SMT constraints produced
+
+1. Write custom rewriting rules that generate constraints symbolically
+    - Pros:
+
+      - Fewer constraints
+      - Higher level of abstraction
+      - Can support multiple output formats
+
+    - Cons:
+
+      - More work
+
+
+## Solution
+
+<!-- Communicates what solution was decided, and it is expected to solve the
+     problem. -->
+
+We propose option (2), and give the following categorization of the Simple fragment:
+
+  - Boolean, integer an uninterpreted literals (including strings)
```

Comment:
```suggestion
  - Boolean, integer and uninterpreted literals (including strings)
```

Reply:
The point of this fragment is that it's basically 1-to-1 with SMT: `T(x /\ y)` becomes `(and T(x) T(y))`, `T( \A x \in Int: P)` becomes `(forall ((x Int)) T(P))`, and so on. There's some room for optimizations with e.g. `\E x \in a..b: P` where you add range memberships as `(and T(P)  T(x <= b) T(x >= a))`, but the point is it's usually direct (no arena tracking, no cherrypick, SMT equality instead of cached axiomatic equality, etc.)